### ISSUE TITLE
Make Structure Refactor

### ISSUE THREAD

**Issue by rlfontenot** [EdwardALuke/loci#20]
Description:
This milestone will make substantial improvements to the Loci build system, pulling in some of the remaining Makefile changes implemented within the CFD Research branch. This milestone will result (probably) in several issues being opened to address the potential work identified below.

Motivation:
The overall goals of this enhancement are to improve the workflow during the build process for Loci developers, to reduce opportunities for common developer errors, and to conform to current best-practices where it makes sense to do so (https://www.gnu.org/software/make/manual/html_node/Makefile-Conventions.html). A secondary goal is to reduce turnaround time for building the code.

Scope:
One example of a developer error to be eliminated within this scope is the version tag which is manually set in lpp/main.cc and src/System/Loci_version.cc. Going forward, the version will be programmatically tied to the git tag system. This change will also need to be trackable within delivered code, but the requisite bundling procedure will be tracked as a separate issue.

The other major class of errors to address are related to cleaning and rebuilding objects. When testing code changes it is desirable to be able to delete and rebuild specific parts of the code as-needed, and the proposed change would enable targeted commands (clean/distclean/build/install) for specific parts of the code (e.g. make distclean_FVMMod; make FVMMod). Additionally, helpful extra information can be communicated to developers during building using text coloring, and before building using make commands for checking the current environment setup.

More fine-grained control over build and install paths is needed. Currently the developer does not have direct control over the build directory location, which can be an issue when developing and testing multiple things in parallel. Adhering to standard use of VPATH should help with this.

Finally, generally following best-practices for Makefiles can help future developers to understand what is going on under the hood. This includes adopting consistent naming conventions for variables.

Requirements:

Implement git tag-based versioning.
Simplify makefiles to reduce opportunities for developer errors.
Use target files to reduce repeated make rules.
Make better use of make features such as filters, implicit rules, and automatic variables to reduce code repetition and manually-maintained lists of objects where possible.
Add "spotless" command for cleaning install directory, and incorporate targeted build/install/clean/distclean/spotless commands.
Add environment variable to control debug builds.
Update to follow "best-practices".
Default to building directly inside the src directory, and allow user to specify their own build directory.
Treat user-specified prefix as the install directory instead of creating a named subdirectory there.
Adhere to common naming conventions, e.g. CPP/CC -> CXX, FC -> F90, LIB_FLAGS -> LDFLAGS, [SHARED/DYNAMIC]_LD -> [SHARED/DYNAMIC]_LD_FLAGS, etc.

**Comment by EdwardALuke** [comment]
It might be a good idea to separate this into a number of separate issues.  Some that I see are:

 * Changing the makefiles to conform to variable standards.  
 * git versioning compatiblilty.  This could tie to some other issues so it might be good to map some things out beforehand.  Some examples of related issues are support for creation of tarball stable releases, but also with the build directory process which I will discuss further below.
 * Implcit make rules are nice, but strike me as a low priority and should probably be relegated to a final cleanup phase.  (Some of these should probably wait until after the FAD updates are more stable, as I can imagine that these will make the make system more complex and make compiling and linking together all files in this directory not the best option.   There might be some approaches using file naming conventions that could help here, but that should be left until we really understand the impacts of the FAD changes.
 * Environment variables to control debug builds can be problematic in some cases, as I mentioned before some debugging options has to be selected at configure time because they are not ABI compatible.  Example when -DBOUNDS_CHECK is set, all files need to be recompiled because some helper class have an additional field to carry array bound information that we don't want to carry in the production compile.  So enabling bounds checking is something that needs to happen at configuration, and not be controlled by environment variables.
 

Aside from this, it is important that some things that the present makefiles do well are preserved.  Two things come to mind :  The first of these is that I make extensive use of parallel make in my development process.  It is important that the Makefiles maximize parallelism and avoid race conditions.  I do most of my development presently on a 128 core EPYC server and routinely remake the entirety of Loci with a compile of about 1 min.   Strategies that reduce available parallelism or introduce races making parallel make unreliable are non-starters regardless of what features are gained in exchange.   Some problems that I saw in the dev branch makesystem in this regard was a serialization of the make flow at the top level makefile, but much more problematically was that subdirectory make files called make in directories above.  Make should only call makefiles in subdirectories.  It is more straightforward to keep the dependencies in sync with the actual workflow in this case, and it makes it much harder to introduce spurious race conditions.  Several times with the dev branch I managed to get make into an infinite loop of makes where one subdirectory called make in another subdirectory only to find it calling a make back to the place that originally called it.  This is why the make calls should form a DAG to prevent this sort of cycle.  In addition, in a parallel make setting it is very easy for such a structure to result in several directories having make running simultaneously that causes race issues.   There will be alot of subtlety to the "default to buiilding directly inside the src directory, so those changes probably need to have a robust set of use cases mapped out ahead of time.  I frequently build sub parts directly in the source directory (For example making development versions of FVMtools or FVMMod using an installed Loci version).  But I suspect that you are after something different with your desire to compile within the src directory.    If there is a desire that the makes within src sub-directories just work, then I think you will be tempted to add rules to recursively make at the upper level, but this has the attendant race condition issues mentioned before.  So I think it is important we get precisely down to the requirements here.

Another issue that was part of the build process frequently used when working on the msstate systems which I would imagine would be extremely useful in other systems as well was the ability to make the build directory exist under the /usr/tmp directory.  This was for several reasons:  1) the source files were usually stored on a NFS mounted, backed up file system which have quotas.   The .o files are large and would often fill users quotas even though these are temporary files.  2) The performance of make when storing and linking .o files from the NFS file system were significantly slower then when these files were stored on the /usr/tmp filesystem (because this is usually a dedicated fast local drive).  Note, that this feature could be problematic with git, as makes performed outside of the git directory will not be able to run git commands to get version information.  We need to retain the ability to compile with the src in a users home directory while the build directory is under /usr/tmp.  The benefits of this are too significant to loose this capability.  

Finally there the Loci install documentation (install.txt) should be updated to document the Loci build process.  But there also needs to be a migration document to help users migrate to the new build system.

**Comment by TimothyEDawson** [comment]
@EdwardALuke I agree that this should probably be split into sub-issues. Raymond and I discussed this earlier today, and we ultimately decided to put it in a single issue mainly to open and track the discussion. We had also looked into some of Github's other options including Milestones (which can be used to group multiple issues) and Projects (Kanban board-like planning), but those both don't allow for comments.

We'll keep looking for good ways to track such discussions, but for now I do expect to ultimately split this into separate issues once we've iterated and agreed upon the requirements.

More later!

**Comment by TimothyEDawson** [comment]
@EdwardALuke While trying to get a handle on what the current "best practices" are for Make, I came across this article which is intriguing: https://accu.org/journals/overload/14/71/miller_2004/

The proposed approach may be outside the scope of this issue, and likely has some downsides, but it may be worth considering as a way to more robustly construct the DAG and thus avoid the race conditions you highlighted. I would be interested in your thoughts.

**Comment by EdwardALuke** [comment]
> @EdwardALuke While trying to get a handle on what the current "best practices" are for Make, I came across this article which is intriguing: https://accu.org/journals/overload/14/71/miller_2004/

I don't think such an approach is necessary for Loci compiles.  The reality is that most dependency problems that happen in Loci makefiles are not due to the recursive makefiles (I will explain why in a bit and also what is the downside (or at least a complication) of the approach listed in the paper).  Generally the only times that I have had problems where I needed to do a make clean to get a reliable make can be traced to one problem.  Right now we don't have automatic dependency analysis for the $include facility implemented by lpp.  Thus if you change a .lh file, make doesn't know about the dependencies that are related to this and usually you need to do a make distclean to make certain that everything that needs to recompile does (unless you manually add the .lh dependencies.)  This has been on my todo list for some time and would solve most dependency problems with current makefiles and has nothing to do with how the makefiles are structured.

The reality is that most of the subdirectories in Loci have very simple dependencies on one another because they are building libraries and not applications.  We can compile both Tools and System at the same time because they don't link to each other even though they utilize each others resources.  Basically the header file dependencies are enough.  The FVMMod FVMOverset and FVMAdapt are just building libraries, but depend on the lpp preprocessor.  So they must be compiled after lpp.  lpp links to Tools so it must be made after libTools is built.  Finally FVMtools must wait for lpp, libTools and libLoci to be built because it links executables that use all of these tools.  So on the top level the dependencies between these modules can be summarized very easily and it is also very unlikely that more complex dependencies than this will develop.  This is the big advantage of the module facility of Loci, it decouples the build process substantially.  If we add more modules to Loci, then the dependencies would be exactly the same as the existing modules if we continue to follow the design idioms that we are presently using.  This also means that we can easily work on one of the modules (say FVMMod) as if it is completely independent of Loci, because it pretty much is.  However, this separation of concerns would have to be carefully built into the approach discussed in the paper.  (It isn't that hard but it would be unfamiliar to many average programmers).  So I don't think this is complicated enough to demand such an approach.  The recursive make system as we have it works well.  The main thing that needs to be fixed is to add a -E type flag to lpp so we can do the $include analysis for the autodependency generation.  That would fix the only real problem that I have seen with the reliability of the present make system, and that has nothing to do with the makefile implementation itself, but rather a missing feature of lpp.

**Comment by EdwardALuke** [comment]
> @EdwardALuke While trying to get a handle on what the current "best practices" are for Make, I came across this article which is intriguing: https://accu.org/journals/overload/14/71/miller_2004/

One advantage of this approach that might be worth considering is that it could dramatically enhance the parallel make potential.  Essentially all files can be converted into object files at the same time, it is only the linking phases that have to be ordered.  The approach described in the paper would allow our make to use much more parallelism.  What we would sacrifice would be the modularity of the make system (e.g. just treating the FVMMod as if it were independent of Loci.)  I will think about this some more and see if we can balance these two issues out.  It might be worth doing an experiment to see how much faster a parallel make can run if we use the approach described in the paper you referenced, but on a 128 core system, I think it would be dramatic.  Probably a low priority issue,but worth thinking about further.

**Comment by EdwardALuke** [comment]
> > @EdwardALuke While trying to get a handle on what the current "best practices" are for Make, I came across this article which is intriguing: https://accu.org/journals/overload/14/71/miller_2004/

For some of the FAD refactoring work I am doing I had to restructure how some of the subdirectories are made in the chem/src directory.  Specifically I needed to link the boundary conditions into the chemCore_FAD module.  The horrible hack I did in the past for the boundary conditions was getting in the way, so I restructured it to use the approach described in the referenced paper.  I had to change the some of the rules in the Loci.conf to make it happen, but it works pretty well.   The changes are in the FAD updates branch (issue #19) that I am currently working on. While this is straightforward for this use case, it does break some of the modularity that you get with the recursive make, namely there are no more make files in the sub directories so it becomes harder to break out a module and work on it independently against a stable version.  But I think we might want to make judicious use of the technique.   It is certainly helpful in managing directories that have large numbers of files to compile that could be further organized into sub-directories to help segregate files into functional units.  I did this in the flowPsi code by naming the files with a prefix that indicated the functional group, but I could see that using directories to do the same thing could result in a much cleaner file layout that may make it easier to find specific functionality.

**Comment by TimothyEDawson** [comment]
@EdwardALuke It does make sense for modules like FVMMod, which are functionally separate from the core Loci code, to have their own self-contained build system. At the same time, as long as it exists inside Loci's source code I think it would be better for everything to use this approach, and we could easily revisit that if it is ever moved elsewhere.

Also, if we set it up with the targeted commands (e.g. make distclean_FVMMod, make FVMMod, make install_FVMMod), that would effectively replace cd'ing into the module directory with commands executed from the root directory. With the added benefit that if you do modify something elsewhere which affects FVMMod, it _should_ be able to detect that and rebuild just what is needed.

**Comment by EdwardALuke** [comment]
> @EdwardALuke It does make sense for modules like FVMMod, which are functionally separate from the core Loci code, to have their own self-contained build system. At the same time, as long as it exists inside Loci's source code I think it would be better for everything to use this approach, and we could easily revisit that if it is ever moved elsewhere.

I am mainly discussing things that are currently part of the process that users are familiar with.  Alot of this has to do with the differences between the stable release and the development version.  Where I see it more significantly is in the CHEM code.  For example, it is not uncommon for me to work with one of the addOn modules that is part of the development, but some of the testing is with regard to the stable release.  For example, the development version may have a significantly reworked particle model that has features that are needed for production work.  It is not uncommon for me to send just the particle model with the make modified so that it generates a module of a different name that can be used to test on production problems where the only thing that changed is that specific part.  These are not going to be built from the top level make, instead they are going to be built from within that module directory with the environment variables pointing to the stable release install.    Now if we ever get to a point where NASA users are willing to use a development version for production work then this issue might not be a big deal, but for that to happen we need to have alot more comprehensive testing infrastructure that we have now.  So I am only highlighting how we are presently using the makefiles.   These processes work right now, so if we are going to change things up, at least we need to be able to support what has worked in the past until we can validate that the new approach can solve the problem in a different way.  In addition we need to have the documentation that will help end users adjust to changing processes.  The present make infrastructure may be non-standard, but it is a sort of defacto standard for current users.  So as we evolve the make system we need to address these issues in the transformation.  If we are going to change the defacto standards then we definitely need to 1) understand how it impacts current workflows and make sure we are discarding an important feature of the current process, and 2) To the extent that we change workflows we need to have clear documentation on how these changes impact currently working systems.  

There are places where the technique can be useful, but we can do both.  I have made changes to the Loci default rules that are compatible with both styles of build system construction, and so far these changes  have had no impact on the existing build infrastructure on flowPsi which I see as a stand in for how intrusive the changes are for end users.
